# 🌻 PRE-ENTREGA:
NOMBRE Y APELLIDO: **CECILIA FONTANA**


---





## 🔆 PRIMERA ETAPA: Recopilación y Preparación de Datos


##🔹1-Crear un documento en Google Colaboratory y cargar los sets de datos como DataFrames

### ✍*IMPORTAR* LAS LIBRERIAS *PANDAS - NUMPY*:

In [1]:
import pandas as pd
import numpy as np

#### *Montar la Unidad

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### *Importar y listar datasets desde Google Drive.


In [3]:
import os
os.listdir("/content/drive/MyDrive/datasets")#Listamos los archivos que contiene la carpeta datasets elegida.

['marketing.csv', 'ventas.csv', 'clientes.csv']

### *Defino rutas relativas de los DataSets

In [4]:
ruta_ventas = "/content/drive/MyDrive/datasets/ventas.csv"
ruta_clientes = "/content/drive/MyDrive/datasets/clientes.csv"
ruta_marketing = "/content/drive/MyDrive/datasets/marketing.csv"

### *Cargo los csv como Data Frames, utilizando el método `read_csv` de Pandas (pd)

In [5]:
ventas = pd.read_csv(ruta_ventas)
clientes = pd.read_csv(ruta_clientes)
marketing = pd.read_csv(ruta_marketing)

### *Valido con un `print` y el método `.shape` para chequear que se cargo correctamente y luego muestro de cada DataFrame, con `display() y .head()` las primeras 4 filas.

In [6]:
print("ventas.shape ->", ventas.shape)
print("clientes.shape ->", clientes.shape)
print("marketing.shape ->", marketing.shape)

print("VENTAS:")
display(ventas.head(4))
print("CLIENTES:")
display(clientes.head(4))
print("MARKETING:")
display(marketing.head(4))

ventas.shape -> (3035, 6)
clientes.shape -> (567, 5)
marketing.shape -> (90, 6)
VENTAS:


id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos

CLIENTES:


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96

MARKETING:


id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024

## 🔹 2- Introducción a Pandas: realizar un análisis exploratorio inicial de los DataFrames.

### *Genero una Función para poder realizar un análisis exploratorio EDA, de cada DataFrame
Utilizo métodos  `.info()`, `.describe()`, `.shape`, `.columns`, `.dtypes.`, `.isna`, `.sum()`


In [7]:
def eda(df, nombre):
    print(f"=== {nombre} ===")
    print("shape:", df.shape)
    print("columnas:", list(df.columns))
    print("dtypes:")
    print(df.dtypes)
    print("\nNulos por columna:")
    print(df.isna().sum())
    print("\nPrimeras filas:")
    display(df.head(5))

    #mostrar un resumen estadístico de las columnas numéricas de un DataFrame (df).
    #El método describe() calcula estadísticas como la media, desviación estándar, mínimo, máximo y percentiles,
    #mientras que include='number' asegura que solo se procesen las columnas numéricas.
    print("\nDescribe (numérico):")
    display(df.describe(include='number'))
    print("-"*100)


### *Ejecuto funcion para DataFrame ventas


In [8]:
eda(ventas, "VENTAS (inicial)")

=== VENTAS (inicial) ===
shape: (3035, 6)
columnas: ['id_venta', 'producto', 'precio', 'cantidad', 'fecha_venta', 'categoria']
dtypes:
id_venta         int64
producto        object
precio          object
cantidad       float64
fecha_venta     object
categoria       object
dtype: object

Nulos por columna:
id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

Primeras filas:


id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos


Describe (numérico):


id_venta     cantidad
count  3035.000000  3033.000000
mean   1499.851400     6.496538
std     866.465379     3.457250
min       1.000000     1.000000
25%     748.500000     3.000000
50%    1502.000000     7.000000
75%    2249.500000     9.000000
max    3000.000000    12.000000

----------------------------------------------------------------------------------------------------


## ✔ Analisis de los datos Iniciales del DataFrame ventas:


*   Tenemos **6** columnas y **3035** registros o filas
*   Hay datos que se deberán normalizar, como por ejemplo **fecha_venta** , **precio**, etc.


*   Tenemos **2 nulos en la columna precio** y **2 en la columna cantidad** que deberemos analizar y eliminar si es preciso.
*   Al describir los datos numéricos contamos con algunas estadísticas que luego nos ayudaran a trabajar con los datos cuando esten normalizados y limpios,





### *Ejecuto funcion para DataFrame clientes


In [9]:
eda(clientes, "CLIENTES (inicial)")

=== CLIENTES (inicial) ===
shape: (567, 5)
columnas: ['id_cliente', 'nombre', 'edad', 'ciudad', 'ingresos']
dtypes:
id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object

Nulos por columna:
id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

Primeras filas:


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96
4           5        Dore Cockshtt    28        Rosario  28245.65


Describe (numérico):


id_cliente        edad      ingresos
count  567.000000  567.000000    567.000000
mean   284.000000   37.940035  34668.739012
std    163.823075   10.202885  12974.531446
min      1.000000   20.000000    170.290000
25%    142.500000   30.000000  26015.240000
50%    284.000000   37.000000  35066.830000
75%    425.500000   43.000000  42457.100000
max    567.000000   81.000000  88053.010000

----------------------------------------------------------------------------------------------------


## ✔ Analisis de los datos Iniciales del DataFrame clientes:


*   Tenemos **5** columnas y **567** registros o filas
*   No se encuentran datos que se deberán normalizar en una primera instancia.

*   No Tenemos valores nulos que se deberían eliminar.
*   Al describir los datos numéricos contamos con algunas estadísticas que luego nos ayudaran a trabajar con los datos cuando esten normalizados y limpios.

### *Ejecuto funcion para DataFrame marketing


In [10]:
eda(marketing, "MARKETING (inicial)")

=== MARKETING (inicial) ===
shape: (90, 6)
columnas: ['id_campanha', 'producto', 'canal', 'costo', 'fecha_inicio', 'fecha_fin']
dtypes:
id_campanha       int64
producto         object
canal            object
costo           float64
fecha_inicio     object
fecha_fin        object
dtype: object

Nulos por columna:
id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

Primeras filas:


id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024


Describe (numérico):


id_campanha      costo
count    90.000000  90.000000
mean     45.500000   4.928667
std      26.124701   0.947750
min       1.000000   2.950000
25%      23.250000   4.372500
50%      45.500000   4.900000
75%      67.750000   5.562500
max      90.000000   7.390000

----------------------------------------------------------------------------------------------------


## ✔ Analisis de los datos Iniciales del DataFrame marketing:


*   Tenemos **6** columnas y **90** registros o filas
*   Se encuentran datos que se deberán normalizar como los de la columna **fecha_inicio**, **fecha_fin**,  en una primera instancia.

*   No Tenemos valores nulos que se deberían eliminar.
*   Al describir los datos numéricos contamos con algunas estadísticas que luego nos ayudaran a trabajar con los datos cuando esten normalizados y limpios.

## 🔹 3- Calidad de Datos (nulos o duplicados)
Identificar valores nulos y duplicados en los conjuntos de datos. Documentar el estado inicial de los datos.

### *Genero una Función para poder realizar un control de la calidad de datos de cada DataFrame y que muestre:


1.   La cantidad de valores nulos por columna
2.   El total de filas completamente duplicadas.


1.  Si se especifica una columna clave, los valores mas repetidos de esa clave.

Utilizo métodos  `.isna`, `.sum()`, `.to_frame`, `.duplicates`, `.value_counts()`, `.sort_values`

In [11]:
def calidad(df, nombre, clave=None):

    print(f"### {nombre}")

    # -------------------------------------------------
    # Mostrar cantidad de valores nulos por columna
    # -------------------------------------------------
    display(df.isna().sum().to_frame("nulos"))

    # -------------------------------------------------
    # Contar filas duplicadas completas
    # -------------------------------------------------

    dup_rows = df.duplicated(keep=False).sum()
    print("Filas duplicadas (exactas):", dup_rows)

    # -------------------------------------------------
    # Si se especificó una columna clave válida, analizar duplicados por esa columna
    # -------------------------------------------------

    if clave and clave in df.columns:
      # Contar cuántas filas tienen valores repetidos en esa columna
        dup_key = df[clave].duplicated(keep=False).sum()
        print(f"Duplicados por clave '{clave}':", dup_key)

        # Si existen duplicados, mostrar cuáles son los valores más repetidos
        if dup_key > 0:

            duplicados_ordenados = (
                df[df[clave].duplicated(keep=False)][clave]
                .value_counts()                # Cuenta cuántas veces aparece cada valor
                .sort_values(ascending=False)   # Ordena de mayor a menor (más duplicados arriba)
            )

            print("\n🔁 Top valores duplicados más frecuentes:")
            # Mostrar solo los primeros 10 (los más repetidos)
            display(duplicados_ordenados.head(10))
        else:
            print(f"No se encontraron duplicados en la clave '{clave}'.")
    else:
        # Si la clave no fue pasada o no existe en el DataFrame
        if clave:
            print(f"La clave '{clave}' no existe en el DataFrame.")
        else:
            print("No se indicó una clave para analizar duplicados por columna.")


### *Ejecuto funcion `calidad`para el DataFrame ventas por su clave `id_venta`


In [12]:
calidad(ventas, "VENTAS", clave="id_venta")


### VENTAS


nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0

Filas duplicadas (exactas): 70
Duplicados por clave 'id_venta': 70

🔁 Top valores duplicados más frecuentes:


id_venta
56      2
421     2
424     2
1868    2
2545    2
2778    2
145     2
300     2
439     2
906     2
Name: count, dtype: int64

## ✔ Analisis de la calidad de los datos DataFrame ventas según su clave.

### *Ejecuto funcion `calidad`para el DataFrame ventas general, sin clave


In [13]:
calidad(ventas, "VENTAS")

### VENTAS


nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0

Filas duplicadas (exactas): 70
No se indicó una clave para analizar duplicados por columna.


## ✔ Analisis de la calidad de los datos DataFrame ventas en general, sin indicar.

### *Ejecuto funcion `calidad`para el DataFrame clientes por su clave `id_cliente`


In [14]:
calidad(clientes, "CLIENTES", clave="id_cliente")


### CLIENTES


nulos
id_cliente      0
nombre          0
edad            0
ciudad          0
ingresos        0

Filas duplicadas (exactas): 0
Duplicados por clave 'id_cliente': 0
No se encontraron duplicados en la clave 'id_cliente'.


## ✔ Analisis de la calidad de los datos DataFrame clientes según su clave.


*   Encontramos que no hay filas duplicadas, según su clave.



### *Ejecuto funcion `calidad`para el DataFrame clientes en general, sin clave.


In [15]:
calidad(clientes, "CLIENTES")


### CLIENTES


nulos
id_cliente      0
nombre          0
edad            0
ciudad          0
ingresos        0

Filas duplicadas (exactas): 0
No se indicó una clave para analizar duplicados por columna.


## ✔ Analisis de la calidad de los datos DataFrame clientes en general, sin indicar.


*   No se encuentran ni nulos, ni filas duplicadas.



### *Ejecuto funcion `calidad`para el DataFrame marketing por su clave `id_campanha`


In [16]:
calidad(marketing, "MARKETING", clave="id_campanha")


### MARKETING


nulos
id_campanha       0
producto          0
canal             0
costo             0
fecha_inicio      0
fecha_fin         0

Filas duplicadas (exactas): 0
Duplicados por clave 'id_campanha': 0
No se encontraron duplicados en la clave 'id_campanha'.


## ✔ Analisis de la calidad de los datos DataFrame marketing en general, indicando su clave.


*   No se encuentran ni nulos, ni filas duplicadas.
Se decide no ejecutar funcion sin la clave, porque el resultado srá el mismo.



## 🔆 SEGUNDA ETAPA: Preprocesamiento y Limpieza de Datos


---





## 🔹 4- Limpieza de Datos: Limpiar el conjunto de datos  eliminando duplicados y caracteres no deseados.
 Documentar el proceso y los resultados.


## **Limpieza del dataset Y normalización del dataset**
- Eliminamos duplicados.
- Normalizamos **texto** en columnas `object` (trim + capitalización simple).
- Convertimos fechas a fechas reales
- Convertimos `precio` y `cantidad` a numéricos si existen.
- Guardamos CSV limpios.

## *Realizamos la  
##### ============================================
#####  🧹 LIMPIEZA Y NORMALIZACIÓN DE LOS DATASETS
##### ============================================
##### Se limpian y normalizan los DataFrames:
#####     ventas, clientes, marketing
##### ============================================

In [17]:
# -------------------------------------------------
# 1️⃣ Crear copias independientes para no modificar los originales
# -------------------------------------------------
ventas_clean = ventas.copy()
clientes_clean = clientes.copy()
marketing_clean = marketing.copy()


In [18]:

# -------------------------------------------------
# 2️⃣ Eliminar filas completamente duplicadas
# -------------------------------------------------
ventas_clean = ventas_clean.drop_duplicates()
clientes_clean = clientes_clean.drop_duplicates()
marketing_clean = marketing_clean.drop_duplicates()

### *Ejecuto funcion `calidad`para los DataFrames copia, luego de haber eliminado las filas completamente duplicadas, para verificar que no queden elementos con èstas caracterìsticas.


In [19]:
calidad(ventas_clean, "VENTAS CLEAN", clave="id_venta")

### VENTAS CLEAN


nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0

Filas duplicadas (exactas): 0
Duplicados por clave 'id_venta': 0
No se encontraron duplicados en la clave 'id_venta'.


Sòlo lo realizamos con el dataset copia clientes, ya que ya hemos detectado con anterioridad que no tenemos duplicados en el los otros dos datasets

### *Genero funcion `normalizar_texto`para limpiar texto en columnas tipo String.


In [24]:
# -------------------------------------------------
# 3️⃣ Función para limpiar texto en columnas tipo string
# -------------------------------------------------
def normalizar_texto(df):
    for col in df.select_dtypes(include="object").columns:

        df[col] = (
            df[col]
            .astype(str)
            .str.strip()
            .str.replace(r"[\u200b\t\r\n]", "", regex=True)
            .str.replace(" +", " ", regex=True)
            .str.title()
        )
    return df

### *Genero algoritmo para normalizar las fechas.


---
Utilizo primero código para normalizar las fechas y luego recién ejecuto
funcion de texto.

In [21]:
# -------------------------------------------------
# Normalizar fechas
# -------------------------------------------------
for df in [ventas_clean, clientes_clean, marketing_clean]:
    for col in df.columns:
        if "fecha" in col.lower():  # detecta columnas con la palabra "fecha"
            df[col] = pd.to_datetime(df[col], errors="coerce", dayfirst=True)


* Corroboro que las fechas quedaron normalizadas

In [22]:
print(ventas_clean.dtypes)
print(clientes_clean.dtypes)
print(marketing_clean.dtypes)

id_venta                int64
producto               object
precio                 object
cantidad              float64
fecha_venta    datetime64[ns]
categoria              object
dtype: object
id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object
id_campanha              int64
producto                object
canal                   object
costo                  float64
fecha_inicio    datetime64[ns]
fecha_fin       datetime64[ns]
dtype: object


### *Ahora corremos la funcion para normalizar textos, e inmediatamente verificamos que asi sea.


In [25]:
#  Aplicar la normalización de texto
# -------------------------------------------------
ventas_clean = normalizar_texto(ventas_clean)
clientes_clean = normalizar_texto(clientes_clean)
marketing_clean = normalizar_texto(marketing_clean)

In [26]:
#mostramos los df luego de normalizar los textos para revisar que queden bien
print(ventas_clean.head(10))
print(clientes_clean.head(10))
print(marketing_clean.head(10))

   id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro Decorativo   $69.94       5.0  2024-01-02   
1       811    Lámpara De Mesa  $105.10       5.0  2024-01-02   
2      1156           Secadora   $97.96       3.0  2024-01-02   
3      1372           Heladera  $114.35       8.0  2024-01-02   
4      1546           Secadora  $106.21       4.0  2024-01-02   
5      1697    Horno Eléctrico   $35.35       9.0  2024-01-02   
6      1710   Plancha De Vapor   $65.43       2.0  2024-01-02   
7      2959          Proyector   $88.17       9.0  2024-01-02   
8       318  Rincón De Plantas   $79.86      11.0  2024-01-03   
9       419         Candelabro   $66.11       8.0  2024-01-03   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos  
5  Electrodomésticos  
6  Electrodomésticos  
7        Electrónica  
8         Decoración  
9         Decoración  
   id_cliente               nom